pkg/rbd/rbd_util.go

```diff
@@ -586,67 +609,43 @@ func deleteSnapshot(ctx context.Context, pOpts *rbdSnapshot, cr *util.Credential
 	if err != nil {
 		return errors.Wrapf(err, "failed to delete snapshot, command output: %s", string(output))
 	}
-
-	if err := undoSnapReservation(ctx, pOpts, cr); err != nil {
-		klog.Errorf(util.Log(ctx, "failed to remove reservation for snapname (%s) with backing snap (%s) on image (%s) (%s)"),
-			pOpts.RequestName, pOpts.RbdSnapName, pOpts.RbdImageName, err)
-	}
-
 	return nil
 }
 
-func restoreSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
+func cloneRbdImageFromSnapshot(ctx context.Context, pVolOpts *rbdVolume, pSnapOpts *rbdSnapshot, cr *util.Credentials) error {
 	var output []byte
 
 	image := pVolOpts.RbdImageName
 	snapName := pSnapOpts.RbdSnapName
 
-	klog.V(4).Infof(util.Log(ctx, "rbd: clone %s using mon %s, pool %s"), image, pVolOpts.Monitors, pVolOpts.Pool)
+	klog.V(4).Infof(util.Log(ctx, "rbd: clone %s/%s@%s %s using mon %s"), pVolOpts.Pool, pSnapOpts.RbdImageName, snapName, image, pVolOpts.Monitors)
 	args := []string{"clone", pSnapOpts.Pool + "/" + pSnapOpts.RbdImageName + "@" + snapName,
-		pVolOpts.Pool + "/" + image, "--id", cr.ID, "-m", pVolOpts.Monitors, "--keyfile=" + cr.KeyFile}
+		pVolOpts.Pool + "/" + image, "--rbd-default-clone-format=2", "--id", cr.ID, "-m", pVolOpts.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature", pVolOpts.ImageFeatures}
 
 	output, err := execCommand("rbd", args)
-
```

Comment:
Nit: perhaps should add flatten depth check here and start a background flatten if required and return `ABORT` error if flatten is needed. Also add in the `CreateSnapshot` path prior to creating the snapshot on the RBD volume image (due to the lack of deep-flatten). Basically, attempt to flatten (if required by depth) both prior to creating any RBD snapshots on an image and after cloning an RBD image.

Reply:
little confused here, are you suggesting to  flatten the source  volume before creating a snapshot and flatten the cloned  image after creating a clone from snapshot during CreateSnapshot

is there a way to find out if the flattening in progress for an image?

Reply:
It's for the case where you have VOL1 (img1) -> SNAP1 (img2) ->VOL2 (img3) -> SNAP2 (img4). We can hit the hard limit at either the snapshot creation or the volume creation.

 If the depth is >= the hard limit, it implies it needs to be flattened and you can schedule it via the background task and `ABORT` so that it backs off. If you fail trying to add the background flatten and need to fall-back to the blocking `rbd flatten`, you could try-lock on affected image and if it is in-progress (lock busy) you can again return `ABORT`

Reply:
@dillaman one concern I have here is, as the volume creation failed with `Abort` error message if the flatten is already in progress as sidecar doesn't send the delete request for the operation in progress(if the PVC delete request is sent), this can lead us to stale backend images

Reply:
What image are you concerned about? The one about to be created (that hasn't been created yet if you need to flatten its parent)?

Reply:
@dillaman parent in the sense cloned volume during snapshot creation, right? if yes how can we handle creating PVC from PVC ( adding background task can leave stale snapshot and clone image) and also just to add here we are already returning `ABORT` error message if  volume operation in progress here https://github.com/ceph/ceph-csi/blob/master/pkg/rbd/controllerserver.go#L138

Reply:
If the CO is free to treat `CreateVolume` `ABORT` errors as if there is nothing to clean up, we have a bigger issue since we could be leaking images all over the place. The CO needs to treat errors from the CSI as an unknown state (maybe the backing PV was created, maybe it wasn't) and force the `DeleteVolume` call to clean up. If this is not happening, I think it's a bug in the CSI spec.

Reply:
yes As  there is no intermediate step between CreateVolume and CreateVolume success, if we return ABORT error message during CreateVolumeRequest as there is no VolumeID is returned if the User sends the PVC delete  request in between as sidecar doesn't have the volumeID to send DeleteRequest the backend image will go stage

Reply:
Right now if you return `ABORT` it will just keep trying [1]. It seems odd that it would allow you to `DeleteVolume` against a PVC that is in-progress. Have you tested it to see if it allows you to do that -- or perhaps it just hangs waiting for `CreateVolume` to succeed (or non-retryable fail) before permitting the `DeleteVolume`.

[1] https://github.com/openshift/csi-external-provisioner/blob/master/pkg/controller/controller.go#L545

Reply:
Did you ever get a chance to clarify if this is a potential for image leaking?

Reply:
verified this once ( will test it a few more times). if the driver creates the rbd image and returns the ABORT error message as flattening in progress. if the PVC is deleted before flattening completes, CreateVolume is not retried till it gets success response nor DeleteVolume will not be called by CSI provisioner as it doesn't have the volume ID (we have an issue on image leaking for smart volume cloning)